### PR TITLE
Add NoUnsetOnPropertyFixer

### LINKED ISSUE #3695: New rule no_unset_on_property
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3692#discussion_r180313322

```diff
 unset($abc)
-unset($this->foo)
+$this->foo = foo;
-unset($this->bar, $this->baz);
+$this->bar = null; $this->baz = null;
```

### PR DESCRIPTION
**Pull request by BackEndTea**
Fixes #3695 

This adds the fixer as described in #3695

### CODE REVIEW

**Comment by keradus** [comment]
Hi !
nice to see this feature started existence !

unsetting key of property is not the same as unsetting the property,
please add test that proves `unset($this->foo[$bar])` is not touch

File: tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php [hunk 1]
```diff
@@ -53,21 +53,28 @@ public function provideFixCases()
...
-                '<?php unset($foo->bar, $bar->foo,   $bar );',
-            ],
-            'It replaces $this->> and self:: replacements' => [
-                '<?php $this->bar = null; self::$foo = null; unset( $bar);',
+            'It replaces $this -> and self:: replacements' => [
+                '<?php $this->bar = null; self::$foo = null; unset($bar);',
                 '<?php unset($this->bar, self::$foo, $bar);',
             ],
+            'It does not replace unsets on arrays' => [
+                '<?php unset($bar->foo[0]);',
```
Comment (keradus) [inline tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php]: add `'<?php unset($bar->foo{0});',` as well.
all hail php...

Comment (BackEndTea) [reply tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php]: I never knew you could do that, and now i wish i didn't know.

Comment (ntzm) [reply tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php]: One thing about writing fixers is you come across a lot of syntax you wish didn't exist 😆

Comment (keradus) [reply tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php]: @BackEndTea , don't you worry. simply start using `normalize_index_brace` and forgot about that syntax in your life ;)
(well, just not in this PR ;P thanks for adding a test :+1:)

File: src/Fixer/LanguageConstruct/NoUnsetOnPropertyFixer.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        return $tokens->isTokenKindFound(T_UNSET)
+            && $tokens->isAnyTokenKindsFound([T_OBJECT_OPERATOR, T_PAAMAYIM_NEKUDOTAYIM]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPriority()
+    {
+        // should be run before CombineConsecutiveUnsetsFixer.
```
Comment (keradus) [inline src/Fixer/LanguageConstruct/NoUnsetOnPropertyFixer.php]: btw,
if it's not a sentence, why final dot is there?

**Comment by BackEndTea** [comment]
I updated doc-blocks, and made sure comments stay preserved.

One edge case that i dont know how to fix is the following:

```diff
- unset/*foo*/($a, $this->a);
+ unset(/*foo*/$a); $this->a = null;
```

If there is a comment between the `unset` and the `(` in a combined unset, where at least one of the unsets needs to be changed to is null, that comment will be moved to after the `(`

**Comment by SpacePossum** [comment]
the move of a bad placed comment to another not so perfect location is not an issue IMHO, as long as we preserve the comment it is up to the end user to fix the mess ;)

**Comment by BackEndTea** [comment]
The Cirlce CI failure looks unrelated to my PR, and i think this fixer is ready.

File: tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php (hunk 1 above)
Comment (SpacePossum) [reply tests/Fixer/LanguageConstruct/NoUnsetOnPropertyFixerTest.php]: you can also mix those up, so please add the test for `<?php unset($bar->foo{0}[1]{2}{$a}[$b]);`

**Comment by keradus** [comment]
while definitely good rule, it is changing code behaviour if one is iterating over instance properties

https://3v4l.org/kiqvB

please mark fixer as risky with damn good description when it's risky.

File: README.rst [hunk 3]
```diff
@@ -1002,6 +1002,10 @@ Choose from the list of available rules:
 
   *Risky rule: modifies the signature of functions; therefore risky when using systems (such as some Symfony components) that rely on those (for example through reflection).*
 
+* **no_unset_on_property**
```
Comment (keradus) [inline README.rst]: please add this rule to local `.php_cs.dist` file

Comment (BackEndTea) [reply README.rst]: Done

**Comment by keradus** [comment]
Thank you @BackEndTea.

**Comment by SpacePossum** [comment]
thank you @BackEndTea , you did all the work here :D  great to see this in!

**Pull request merged**